File: stitch_agent/run/runner.py
"""Runner loop -- orchestrate local CI execution with an AI fix loop.

Jobs are executed in parallel via asyncio.gather. When multiple jobs fail,
their errors are batched into a single driver prompt to minimize LLM calls.
After each batch fix, all previously-failed jobs are re-run in parallel to
check which ones the fix resolved.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import TYPE_CHECKING, Protocol

from stitch_agent.run.drivers.base import build_batch_prompt
from stitch_agent.run.executor import LocalExecutor
from stitch_agent.run.models import (
    CIJob,
    FixContext,
    JobResult,
    RunReport,
)

if TYPE_CHECKING:
    from pathlib import Path

    from stitch_agent.run.drivers.base import AgentDriver
    from stitch_agent.run.executor import ExecResult

_ERROR_LOG_TAIL_CHARS = 4_000


class RunnerCallback(Protocol):
    """Callback protocol for UI integration."""

    def job_started(self, name: str, attempt: int, max_attempts: int) -> None: ...
    def job_log_update(self, name: str, log: str) -> None: ...
    def job_finished(self, name: str, result: JobResult) -> None: ...
    def driver_started(self, name: str, driver_name: str) -> None: ...
    def driver_log_update(self, name: str, log: str) -> None: ...


class _NullCallback:
    def job_started(self, name: str, attempt: int, max_attempts: int) -> None:
        pass

    def job_log_update(self, name: str, log: str) -> None:
        pass

    def job_finished(self, name: str, result: JobResult) -> None:
        pass

    def driver_started(self, name: str, driver_name: str) -> None:
        pass

    def driver_log_update(self, name: str, log: str) -> None:
        pass


@dataclass
class RunnerConfig:
    max_attempts: int = 3
    fail_fast: bool = False
    job_timeout_seconds: float = 300.0


class Runner:
    def __init__(
        self,
        repo_root: Path,
        driver: AgentDriver,
        config: RunnerConfig | None = None,
        executor: LocalExecutor | None = None,
        callback: RunnerCallback | None = None,
    ) -> None:
        self.repo_root = repo_root
        self.driver = driver
        self.config = config or RunnerConfig()
        self.executor = executor or LocalExecutor(
            repo_root, timeout_seconds=self.config.job_timeout_seconds,
        )
        self._cb: RunnerCallback = callback or _NullCallback()

    async def run(
        self, jobs: list[CIJob], dry_run: bool = False,
    ) -> RunReport:
        results: dict[str, JobResult] = {}
        runnable: list[CIJob] = []

        for job in jobs:
            if job.skip_reason:
                results[job.name] = JobResult(
                    name=job.name,
                    status="skipped",
                    skip_reason=job.skip_reason,
                )
            elif dry_run:
                results[job.name] = JobResult(name=job.name, status="not_run")
            else:
                runnable.append(job)

        if not runnable:
            return RunReport(
                jobs=[results[j.name] for j in jobs],
                agent=self.driver.name,
            )

        pending = list(runnable)

        for attempt in range(1, self.config.max_attempts + 1):
            # Run all pending jobs in parallel
            exec_results = await self._run_jobs_parallel(pending, attempt)

            # Partition into passed and failed
            failed: list[tuple[CIJob, str]] = []
            for job in pending:
                er = exec_results[job.name]
                if er.exit_code == 0:
                    result = JobResult(
                        name=job.name, status="passed", attempts=attempt,
                    )
                    results[job.name] = result
                    self._cb.job_finished(job.name, result)
                else:
                    failed.append((job, er.log))

            if not failed:
                break

            # Last attempt: escalate all remaining failures
            if attempt >= self.config.max_attempts:
                for job, log in failed:
                    result = JobResult(
                        name=job.name,
                        status="escalated",
                        attempts=attempt,
                        driver=self.driver.name,
                        error_log=log[-_ERROR_LOG_TAIL_CHARS:],
                    )
                    results[job.name] = result
                    self._cb.job_finished(job.name, result)
                break

            # Batch fix all failures in a single driver call
            contexts = [
                FixContext(
                    repo_root=self.repo_root,
                    job_name=job.name,
                    command=" && ".join(job.script),
                    script=list(job.script),
                    error_log=log,
                    attempt=attempt,
                )
                for job, log in failed
            ]

            batch_context = self._make_batch_context(contexts)
            batch_label = ", ".join(j.name for j, _ in failed)

            self._cb.driver_started(batch_label, self.driver.name)

            self.driver.on_output = lambda log, _label=batch_label: (
                self._cb.driver_log_update(_label, log)
            )

            outcome = await self.driver.fix(batch_context)
            self.driver.on_output = None

            if not outcome.applied:
                reason = outcome.reason or "driver did not apply a fix"
                for job, log in failed:
                    result = JobResult(
                        name=job.name,
                        status="escalated",
                        attempts=attempt,
                        driver=self.driver.name,
                        error_log=(
                            f"[driver: {reason}]\n\n"
                            + log[-_ERROR_LOG_TAIL_CHARS:]
                        ),
                    )
                    results[job.name] = result
                    self._cb.job_finished(job.name, result)
                break

            # Next round: re-run only the jobs that failed
            pending = [job for job, _ in failed]

            if self.config.fail_fast:
                # In parallel model, fail_fast halts after escalation
                # (max_attempts or driver refusal), both handled above.
                pass

        # Build report preserving original job order
        ordered = [
            results.get(j.name, JobResult(name=j.name, status="not_run"))
            for j in jobs
        ]
        return RunReport(jobs=ordered, agent=self.driver.name)

    async def _run_jobs_parallel(
        self, jobs: list[CIJob], attempt: int,
    ) -> dict[str, ExecResult]:
        """Run multiple jobs concurrently, return results keyed by name."""
        for job in jobs:
            self._cb.job_started(job.name, attempt, self.config.max_attempts)

        async def run_one(job: CIJob) -> tuple[str, ExecResult]:
            result = await self.executor.run_job(job)
            self._cb.job_log_update(job.name, result.log)
            return job.name, result

        pairs = await asyncio.gather(*(run_one(j) for j in jobs))
        return dict(pairs)

    def _make_batch_context(
        self, contexts: list[FixContext],
    ) -> FixContext:
        """Merge multiple FixContexts into one with a batch prompt."""
        if len(contexts) == 1:
            return contexts[0]

        prompt = build_batch_prompt(contexts)
        return FixContext(
            repo_root=contexts[0].repo_root,
            job_name=", ".join(c.job_name for c in contexts),
            command="(batch fix)",
            script=[],
            error_log="",
            attempt=contexts[0].attempt,
            prompt_override=prompt,
        )

File: src/core/runner.ts
import { buildBatchPrompt } from "../drivers/prompt.js";
import type { AgentDriver } from "../drivers/types.js";
import { LocalExecutor } from "./executor.js";
import type { CIJob, ExecResult, FixContext, JobResult } from "./models.js";
import { RunReport } from "./models.js";

interface JobExecutor {
  runJob(job: CIJob, signal?: AbortSignal): Promise<ExecResult>;
}

const ERROR_LOG_TAIL_CHARS = 4_000;

export interface RunnerCallback {
  jobStarted(name: string, attempt: number, maxAttempts: number): void;
  jobLogUpdate(name: string, log: string): void;
  jobFinished(name: string, result: JobResult): void;
  driverStarted(name: string, driverName: string): void;
  driverLogUpdate(name: string, log: string): void;
}

class NullCallback implements RunnerCallback {
  jobStarted(_name: string, _attempt: number, _maxAttempts: number): void {}
  jobLogUpdate(_name: string, _log: string): void {}
  jobFinished(_name: string, _result: JobResult): void {}
  driverStarted(_name: string, _driverName: string): void {}
  driverLogUpdate(_name: string, _log: string): void {}
}

export interface RunnerConfig {
  maxAttempts: number;
  failFast: boolean;
  jobTimeoutSeconds: number;
}

const DEFAULT_CONFIG: RunnerConfig = {
  maxAttempts: 3,
  failFast: false,
  jobTimeoutSeconds: 300,
};

function notRunResult(name: string): JobResult {
  return {
    name,
    status: "not_run",
    attempts: 0,
    driver: null,
    errorLog: "",
    skipReason: null,
  };
}

export class Runner {
  private repoRoot: string;
  private driver: AgentDriver;
  private config: RunnerConfig;
  private executor: JobExecutor;
  private cb: RunnerCallback;

  constructor(
    repoRoot: string,
    driver: AgentDriver,
    config?: Partial<RunnerConfig>,
    executor?: JobExecutor,
    callback?: RunnerCallback,
  ) {
    this.repoRoot = repoRoot;
    this.driver = driver;
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.executor = executor ?? new LocalExecutor(repoRoot, this.config.jobTimeoutSeconds);
    this.cb = callback ?? new NullCallback();
  }

  async run(jobs: CIJob[], dryRun = false, signal?: AbortSignal): Promise<RunReport> {
    const results = new Map<string, JobResult>();
    const runnable = this.partitionJobs(jobs, dryRun, results);

    if (runnable.length !== 0) {
      await this.runAttempts(runnable, results, signal);
    }

    const ordered = jobs.map((j) => results.get(j.name) ?? notRunResult(j.name));
    return new RunReport(ordered, this.driver.name);
  }

  private async runAttempts(
    runnable: CIJob[],
    results: Map<string, JobResult>,
    signal?: AbortSignal,
  ): Promise<void> {
    let pending = [...runnable];
    for (let attempt = 1; attempt <= this.config.maxAttempts; attempt++) {
      if (signal?.aborted) {
        this.markAborted(pending, attempt, results);
        return;
      }

      const useFailFast = this.config.failFast && attempt < this.config.maxAttempts;
      const execResults = await this.runJobsParallel(pending, attempt, useFailFast, signal);
      if (signal?.aborted) {
        this.markAborted(pending, attempt, results);
        return;
      }

      const { failed, cancelled } = this.classifyResults(pending, execResults, attempt, results);
      if (failed.length === 0 && cancelled.length === 0) return;
      if (failed.length === 0) {
        pending = cancelled;
        continue;
      }

      if (attempt >= this.config.maxAttempts) {
        this.escalateJobs(failed, attempt, results);
        return;
      }

      const nextPending = await this.runFixStep(failed, cancelled, attempt, results, signal);
      if (nextPending === null) return;
      pending = nextPending;
    }
  }

  private async runFixStep(
    failed: [CIJob, string][],
    cancelled: CIJob[],
    attempt: number,
    results: Map<string, JobResult>,
    signal?: AbortSignal,
  ): Promise<CIJob[] | null> {
    const contexts: FixContext[] = failed.map(([job, log]) => ({
      repoRoot: this.repoRoot,
      jobName: job.name,
      command: job.script.join(" && "),
      script: [...job.script],
      errorLog: log,
      attempt,
      promptOverride: null,
    }));

    const batchContext = this.makeBatchContext(contexts);
    const batchLabel = failed.map(([j]) => j.name).join(", ");

    this.cb.driverStarted(batchLabel, this.driver.name);
    this.driver.onOutput = (log: string) => this.cb.driverLogUpdate(batchLabel, log);
    const outcome = await this.driver.fix(batchContext, signal);
    this.driver.onOutput = null;

    if (signal?.aborted) {
      this.markAborted([...failed.map(([job]) => job), ...cancelled], attempt, results);
      return null;
    }
    if (!outcome.applied) {
      this.escalateJobs(failed, attempt, results, outcome.reason);
      return null;
    }
    return [...failed.map(([job]) => job), ...cancelled];
  }

  private markAborted(jobs: CIJob[], attempt: number, results: Map<string, JobResult>): void {
    for (const job of jobs) {
      if (results.has(job.name)) continue;
      const result: JobResult = {
        name: job.name,
        status: "not_run",
        attempts: attempt,
        driver: null,
        errorLog: "",
        skipReason: "aborted",
      };
      results.set(job.name, result);
      this.cb.jobFinished(job.name, result);
    }
  }

  private async runJobsParallel(
    jobs: CIJob[],
    attempt: number,
    failFast: boolean,
    external?: AbortSignal,
  ): Promise<Map<string, ExecResult>> {
    for (const job of jobs) {
      this.cb.jobStarted(job.name, attempt, this.config.maxAttempts);
    }

    const controller = failFast || external ? new AbortController() : undefined;
    const onExternalAbort = () => controller?.abort();
    if (external && controller) {
      if (external.aborted) controller.abort();
      else external.addEventListener("abort", onExternalAbort, { once: true });
    }

    const pairs = await Promise.all(
      jobs.map(async (job): Promise<[string, ExecResult]> => {
        const result = await this.executor.runJob(job, controller?.signal);
        if (!result.cancelled) {
          this.cb.jobLogUpdate(job.name, result.log);
        }
        if (failFast && !result.cancelled && result.exitCode !== 0) {
          controller?.abort();
        }
        return [job.name, result];
      }),
    );

    external?.removeEventListener("abort", onExternalAbort);

    return new Map(pairs);
  }

  private classifyResults(
    pending: CIJob[],
    execResults: Map<string, ExecResult>,
    attempt: number,
    results: Map<string, JobResult>,
  ): { failed: [CIJob, string][]; cancelled: CIJob[] } {
    const failed: [CIJob, string][] = [];
    const cancelled: CIJob[] = [];
    for (const job of pending) {
      const er = execResults.get(job.name);
      if (!er) continue;
      if (er.cancelled) {
        cancelled.push(job);
        continue;
      }
      if (er.exitCode === 0) {
        const result: JobResult = {
          name: job.name,
          status: "passed",
          attempts: attempt,
          driver: null,
          errorLog: "",
          skipReason: null,
        };
        results.set(job.name, result);
        this.cb.jobFinished(job.name, result);
      } else {
        failed.push([job, er.log]);
      }
    }
    return { failed, cancelled };
  }

  private escalateJobs(
    failed: [CIJob, string][],
    attempt: number,
    results: Map<string, JobResult>,
    driverReason?: string,
  ): void {
    for (const [job, log] of failed) {
      const prefix = driverReason ? `[driver: ${driverReason}]\n\n` : "";
      const result: JobResult = {
        name: job.name,
        status: "escalated",
        attempts: attempt,
        driver: this.driver.name,
        errorLog: `${prefix}${log.slice(-ERROR_LOG_TAIL_CHARS)}`,
        skipReason: null,
      };
      results.set(job.name, result);
      this.cb.jobFinished(job.name, result);
    }
  }

  private partitionJobs(jobs: CIJob[], dryRun: boolean, results: Map<string, JobResult>): CIJob[] {
    const runnable: CIJob[] = [];
    for (const job of jobs) {
      if (job.skipReason) {
        results.set(job.name, {
          name: job.name,
          status: "skipped",
          attempts: 0,
          driver: null,
          errorLog: "",
          skipReason: job.skipReason,
        });
      } else if (dryRun) {
        results.set(job.name, notRunResult(job.name));
      } else {
        runnable.push(job);
      }
    }
    return runnable;
  }

  private makeBatchContext(contexts: FixContext[]): FixContext {
    const [first, ...rest] = contexts;
    if (!first) throw new Error("makeBatchContext called with empty contexts");
    if (rest.length === 0) return first;

    const prompt = buildBatchPrompt(contexts);
    return {
      repoRoot: first.repoRoot,
      jobName: contexts.map((c) => c.jobName).join(", "),
      command: "(batch fix)",
      script: [],
      errorLog: "",
      attempt: first.attempt,
      promptOverride: prompt,
    };
  }
}
